packages/matter.js/src/ble/BtpSessionHandler.ts

```diff
@@ -0,0 +1,423 @@
+                }
+                if (this.currentIncomingSegmentedPayload.length !== this.currentIncomingSegmentedMsgLength) {
+                    throw new BtpProtocolError(`BTP packet payload length does not match message length: ${this.currentIncomingSegmentedPayload.length} !== ${this.currentIncomingSegmentedMsgLength}`);
+                }
+
+                const payloadToProcess = this.currentIncomingSegmentedPayload;
+                this.currentIncomingSegmentedMsgLength = undefined
+                this.currentIncomingSegmentedPayload = undefined; // resetting current segment Payload to empty byte array
+
+                // Hand over the resulting Matter message to ExchangeManager via the callback
+                await this.handleMatterMessagePayload(payloadToProcess);
+            }
+
+        } catch (error) {
+            logger.error(`Error while handling incoming BTP data: ${error}`);
+            await this.close();
+            if (!(error instanceof BtpProtocolError)) { // If no BTP protocol error, rethrow
+                throw error;
+            }
+        }
+    }
+
+    /**
+     * Send a Matter message to the transport layer, but before that encode it into a BTP packet and potentially split
+     * it into multiple segments. This Method is indirectly called by the ExchangeManager layer when a Matter message
+     * should be sent.
+     *
+     * @param data ByteArray containing the Matter message
+     */
+    public async sendMatterMessage(data: ByteArray) {
+        if (!this.isActive) {
+            throw new BtpFlowError("BTP session is not active");
+        }
+        logger.debug(`Got Matter message to send via BLE transport: ${data.toHex()}`);
+
+        if (data.length === 0) {
+            throw new BtpFlowError("BTP packet must not be empty");
+        }
+        const dataReader = new DataReader(data, Endian.Little);
+        this.queuedOutgoingMatterMessages.push(dataReader);
+        await this.processSendQueue();
+    }
+
+    private async processSendQueue() {
+        if (this.sendInProgress) return;
+
+        if (this.exceedsWindowSize(this.prevIncomingAckNumber, this.sequenceNumber)) return;
+
+        if (this.queuedOutgoingMatterMessages.length === 0) return;
+
+        this.sendInProgress = true;
+
+        while (this.queuedOutgoingMatterMessages.length > 0) {
+            const currentProcessedMessage = this.queuedOutgoingMatterMessages[0];
+            const remainingMessageLength = currentProcessedMessage.getRemainingBytesCount();
+
+            logger.debug(`Sending BTP fragment: ${Logger.dict({
+                fullMessageLength: currentProcessedMessage.getLength(),
+                remainingLengthInBytes: remainingMessageLength,
+            })}`);
+
+            //checks if last ack number sent < ack number to be sent
+            const hasAckNumber = this.prevIncomingSequenceNumber !== this.prevAckedSequenceNumber;
```

Comment:
what if `this.prevIncomingSequenceNumber < this.prevAckedSequenceNumber` ?? 
Then the hasAckNumber should not be true, right?
It should ideally throw an error?

Reply:
if the client window is 255 then such a case could happen. We have many places to check that all data is right when packates come in and such, so I decided to trust here that the data are correct and just send out an ack number if the current data are not "the same".
Alternatively we also need to add a "exceedsWindowSize" check here to throw. But before we do that we should check that this case can really happen. i guess not